hnn_core/opt_toy_example/general.py

```diff
@@ -0,0 +1,378 @@
+        drive_predicted_params.update({'mu': params[f'{drive_name}_mu']})
+        drive_predicted_params.update({'sigma': params[f'{drive_name}_sigma']})
+        ampa_weights = dict()
+        nmda_weights = dict()
+        conn_idxs = pick_connection(net, src_gids=drive_name)
+        for conn_idx in conn_idxs:
+            target_type = net.connectivity[conn_idx]['target_type']
+            target_receptor = net.connectivity[conn_idx]['receptor']
+            if target_receptor == 'ampa':
+                ampa_weights.update(
+                    {target_type: params[f'{drive_name}_ampa_{target_type}']})
+            elif target_receptor == 'nmda':
+                nmda_weights.update(
+                    {target_type: params[f'{drive_name}_nmda_{target_type}']})
+        drive_predicted_params.update({'ampa_weights': ampa_weights})
+        drive_predicted_params.update({'nmda_weights': nmda_weights})
+        predicted_params.update({drive_name: drive_predicted_params})
+    return predicted_params
+
+
+def _set_params(net, fixed_params, predicted_params):
+    """Sets the network parameters.
+
+       Parameters
+       ----------
+       net : the Network object
+       fixed_params : unchanging network parameters
+       predicted_params : the parameters predicted by the optimizer
+
+       Returns
+       -------
+       net : Network object
+    """
+
+    net_new = net.copy()
+    # remove existing drives, re-set them with updated parameters
+    for drive_name in net.external_drives.keys():
+        # clear drive
+        del net_new.external_drives[drive_name]
+        # clear connectivity
+        conn_idxs = pick_connection(net_new, src_gids=drive_name)
+        net_new.connectivity = [conn for conn_idx, conn
+                                in enumerate(net_new.connectivity)
+                                if conn_idx not in conn_idxs]
+        net_new.add_evoked_drive(name=drive_name,
+                                 mu=predicted_params[drive_name]['mu'],
+                                 sigma=predicted_params[drive_name]['sigma'],
+                                 numspikes=fixed_params[drive_name]
+                                 ['numspikes'],
+                                 location=fixed_params[drive_name]['location'],
+                                 n_drive_cells=fixed_params[drive_name]
+                                 ['n_drive_cells'],
+                                 cell_specific=fixed_params[drive_name]
+                                 ['cell_specific'],
+                                 weights_ampa=predicted_params[drive_name]
+                                 ['ampa_weights'],
+                                 weights_nmda=predicted_params[drive_name]
+                                 ['nmda_weights'],
+                                 space_constant=fixed_params[drive_name]
+                                 ['space_constant'],
+                                 synaptic_delays=fixed_params[drive_name]
```

Comment:
I think the more generalizable way is to update the network attributes directly ... basically do the opposite of what you did in `_get_params`

Reply:
Shouldn't we strive to use the established code path via the public API since setting `Network` attributes can be much trickier than getting them? Mostly just because there are a bunch of checks and stuff in `Network.add_evoked_drive()` and `Network._attach_drive()`. For instance, if the user should decide to optimize the `n_drive_cells` parameter (not sure why one would want to do this, but it's theoretically an option), we'd need to make sure the `gid_ranges` and `connectivity` attributes are updated accordingly.

Reply:
but that seemed to be different from what [you proposed here](https://github.com/jonescompneurolab/hnn-core/pull/652#discussion_r1232398759) ?

Reply:
I meant that the general strategy is to retrieve fixed parameters from `Network` attributes that should be treated more-or-less as immutable (e.g., like an external drive), and then re-instantiate them using the standard API with their original fixed params and updated params. If the `Network` attribute getting modified during optimization is fair-game for direct modification, the `_get_params()` method is irrelevant (right?). Some `Network` attributes can be overwritten directly (particularly those without a complex instantiation API written for them), but others are a bit error prone.

Reply:
> In that case, maybe it's better if we initially restrict the params to be amongst those that can be instantiated through the API. If we assemble the dictionary cleverly, we might even be able to just say:

You mean, restrict optimization params to just drives? Or drives and connectivity maybe?

> net.add_evoked_drive(**updated_params)
> 
> Where updated_params are the arguments relevant for the evoked drive. You can get the list of arguments a function accepts and make the dictionary keys match those.

Reply:
Start with drives and drive connectivity. We can always expand to other parameters.